app/celer-sim/celer-sim.cc

```diff
@@ -159,121 +165,88 @@ std::shared_ptr<OutputRegistry> run(std::istream* is)
 
-//---------------------------------------------------------------------------//
-/*!
- * Execute and run.
- */
 int main(int argc, char* argv[])
 {
-    using celeritas::ScopedMpiInit;
-    using celeritas::to_string;
+    using namespace celeritas::app;
     using std::cout;
     using std::endl;
 
-    ScopedMpiInit scoped_mpi(&argc, &argv);
+    // Set up MPI
+    celeritas::ScopedMpiInit scoped_mpi(&argc, &argv);
     if (scoped_mpi.is_world_multiprocess())
     {
         CELER_LOG(critical) << "TODO: this app cannot run in parallel";
         return EXIT_FAILURE;
     }
 
-    // Process input arguments
-    if (argc != 2)
-    {
-        celeritas::app::print_usage(argv[0]);
-        return EXIT_FAILURE;
-    }
-    std::string_view filename{argv[1]};
-    if (filename == "--help"sv || filename == "-h"sv)
-    {
-        celeritas::app::print_usage(argv[0]);
-        return EXIT_SUCCESS;
-    }
-    if (filename == "--version"sv || filename == "-v"sv)
-    {
-        std::cout << celeritas::version_string << std::endl;
-        return EXIT_SUCCESS;
-    }
-    if (filename == "--config"sv)
-    {
-        std::cout << to_string(celeritas::BuildOutput{}) << std::endl;
-        return EXIT_SUCCESS;
-    }
-    if (filename == "--dump-default"sv)
-    {
-        std::cout << nlohmann::json(celeritas::app::RunnerInput{}).dump(1)
-                  << std::endl;
-        return EXIT_SUCCESS;
-    }
+    // Set up app
+    CLI::App cli{"Run standalone Celeritas"};
+    setup_app(cli);
 
-    // Initialize GPU
-    if (filename == "--device"sv)
-    {
-        celeritas::activate_device();
+    std::string filename;
+    cli.add_option("filename", filename, "Input JSON")
+        ->check(CLI::ExistingFile | dash_validator());
 
-        if (celeritas::Device::num_devices() == 0)
-        {
-            CELER_LOG(critical) << "No GPUs were detected";
-            return EXIT_FAILURE;
-        }
-        std::cout << nlohmann::json(celeritas::device()).dump(1) << std::endl;
-        return EXIT_SUCCESS;
-    }
+    std::function<std::string()> diagnostic;
+    auto set_diagnostic = [&diagnostic](auto func) {
+        return [&diagnostic, func = std::move(func)](auto count) {
+            CELER_DISCARD(count);
+            diagnostic = std::move(func);
+        };
+    };
+    cli.add_flag(
+        "--config",
+        set_diagnostic([] { return to_string(celeritas::BuildOutput{}); }),
+        "Show configuration");
+    cli.add_flag("--dump-default",
+                 set_diagnostic(get_default_string),
+                 "Dump default input");
+    cli.add_flag("--device",
+                 set_diagnostic(get_device_string),
+                 "Show device information");
 
-    std::ifstream infile;
-    std::istream* instream = nullptr;
-    if (filename == "-")
-    {
-        instream = &std::cin;
-        filename = "<stdin>";  // For nicer output on failure
-    }
-    else
+    // Parse and run
+    CELER_CLI11_PARSE(cli, argc, argv);
```

Comment:
I guess since there are no required options we don't see the parse error and usage printed when running without any arguments, just a runtime error? Should we require that either the input json or one of the flags be provided?

Reply:
Fixed.